npep/npep-122.md

```diff
@@ -0,0 +1,181 @@
+
+**User story**
+
+Traffic should be disallowed by default between user namespaces owned by different users.
+
+**Policy**
+```
+kind: BaselineAdminNetworkPolicy
+apiVersion: policy.networking.k8s.io/v1alpha1
+metadata:
+  name: user-based-tenancy
+spec:
+  subject:
+    namespaces:
+      matchExpressions:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces:
+            notSameLabels:
+              - user
+```
+
+**Meaning**
+
+In namespaces that have a “user” label, by default, deny ingress from namespaces that have a different “user” label.
+
+There are two problems with this implementation of tenancy, relative to the user stories that it exists to support
+
+First, it never explicitly defines what a tenant is. Although the object correctly implements the desired tenancy 
+policy, it doesn’t really make it obvious that that’s what it’s doing.
+
+Second, the syntax is very general purpose / powerful. ANP has subjects and peers, which are different, 
+and currently Tenancy is defined on the peers side. Tenancy by itself has the same subject and peer, 
+at least for the existing use cases, and having separate selectors for subject and peer allows for more
+configurations than needed.
+
+## API
+
+We suggest to define a separate API for tenancy, as in the following example:
+
+```
+kind: AdminNetworkPolicyTenancy
+apiVersion: policy.networking.k8s.io/v1alpha1
+metadata:
+  name: user-based
+spec:
+  namespaceSelector:
+    matchExpression:
+      - key: system-namespace
+        operator: DoesNotExist
+      - key: user
+        operator: Exists
+  tenancyLabels:
+    - user
+  priority: 99
+  ingress:
+    - action: Deny
+      from:
+        - NotSameTenant
+```
+
+that will define a tenant as follows:
+- first, select namespaces that will be split into tenants, and have the defined policy applied. `spec.namespaceSelector`
+is used for this purpose. This also solves the problem we had for empty or unset labels: if tenancy is defined based on label `user`,
+and namespaces without this label should not be affected, `key: user, operator: Exists` selector can be used.
+- second, selected namespaces are split into tenants based on label values, `spec.tenancyLabels` field is used for that purpose.
+Tenants with the same listed label values will form a tenant. Unset label and empty string label are also values, therefore
+in case namespaces with these label values are selected, they will be split into a separate tenant.
+
+    > Example: say we define a tenancy based on `label1` and `label2`, and the following namespaces are selected by `spec.namespaceSelector`
+    > ```
+    > ns1{label1=A, label2=B, label3=C}
+    > ns2{label1=A, label2=B, label4=D}
+    > ns3{label1=_, label2=B}
+    > ns4{label1=_, label2=""}
+    > ns5{label1=_, label2=_}
+    > ns6{label1=A, label2=B}
+    > ```
+    > where `_` means the label is not set, and `""` means empty string value.
```

Comment:
We can touch upon naming later but I just what to point out that, this categorization of tenants is quite different than our original `sameLabels` proposal, where namespaces who do not have the tenancy label at all (as opposed to have an empty label) are not affected by the policy. 
When I read the `tenancyLabels` field, it is supposed to mean "namespaces that have the exact same values for these labels is considered a single tenant", but it does not and should not IMO imply that namespaces that are missing this label automatically form a tenant by themselves

Reply:
Current design is supposed to be more explicit on which namespaces should be affected, and also allow different configurations. So if you don't want namespace without a "tenancy label" to be affected, you just add
`key: <tenancy label>, operator: Exists` to the `spec.namespaceSelector`. But at the same time if you do want to make a separate tenant for all namespaces without that label you leave empty selector.

Also I think `tenancyLabels` does exactly what you said: "namespaces that have the exact same values for these labels are considered a single tenant", because all namespaces with empty label will be one tenant, and all namespaces without a label (say they all have None value for it) will be one tenant. And you can easily avoid namespaces with unset labels as I pointed above.

Reply:
We could alternatively say that namespaces without all of the tenancyLabels are always excluded from the definition. IOW, `namespaceSelector` automatically implicitly includes `key: X, operator: Exists` for every label in `tenancyLabels`

Reply:
We can do that, but I see 2 possible disadvantages to this:
- `namespaceSelector` doesn't fully define which namespaces will be selected, so we add an implicit label selector `key: X, operator: Exists` and it makes the API more confusing
- there is no way to set rules for namespaces with label unset: e.g. if I want to deny-from-not-same-tenant, I may want to deny from not only namespaces with different label value, but also from namespaces without a label, otherwise forgetting to set a label on a namespace may become a security hole, because it won't be affected by tenancy.